### ISSUE TITLE
---> 46 playwright = AsyncPlaywright(next(iter(done)).result()) [BUG]

### ISSUE THREAD

**Issue by aford4141** [microsoft/playwright-python#1710]
Hello 

Here is the code that i am working on and I keep getting this NotImplementedError: from the playwrite. Can someone help!! 

import os
from bs4 import BeautifulSoup
from playwright.async_api import async_playwright, TimeoutError as PlaywrightTimeout
import time
%pip install pandas
Requirement already satisfied: pandas in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (1.5.2)
Requirement already satisfied: python-dateutil>=2.8.1 in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (from pandas) (2.8.2)
Requirement already satisfied: pytz>=2020.1 in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (from pandas) (2022.7)
Requirement already satisfied: numpy>=1.21.0 in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (from pandas) (1.24.1)
Requirement already satisfied: six>=1.5 in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (from python-dateutil>=2.8.1->pandas) (1.16.0)
Note: you may need to restart the kernel to use updated packages.
%pip install beautifulsoup4
Requirement already satisfied: beautifulsoup4 in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (4.11.1)
Requirement already satisfied: soupsieve>1.2 in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (from beautifulsoup4) (2.3.2.post1)
Note: you may need to restart the kernel to use updated packages.
%pip install playwright
Requirement already satisfied: playwright in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (1.29.1)
Requirement already satisfied: greenlet==2.0.1 in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (from playwright) (2.0.1)
Requirement already satisfied: pyee==9.0.4 in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (from playwright) (9.0.4)
Requirement already satisfied: typing-extensions in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (from pyee==9.0.4->playwright) (4.4.0)
Note: you may need to restart the kernel to use updated packages.
!playwright install
SEASONS = list(range(2016, 2023))
SEASONS
[2016, 2017, 2018, 2019, 2020, 2021, 2022]
DATA_DIR = "data"
STANDINGS_DIR = os.path.join(DATA_DIR, "standings")
SCORES_DIR = os.path.join(DATA_DIR, "scores")
async def get_html(url, selector, sleep=5, retries=5):
    html = None
    for i in range(1, retries+1):
        time.sleep(sleep * i)
        
        try:
            async with async_playwright() as p:
                browser = await p.chromium.launch()
                page= await browser.new_page()
                await page.goto(url)
                print(await page.title())
                html = await page.inner_html(selector)
        except PlaywrightTimeout:
            print(f"Timeout error on {url}")
            continue
        else:
            break
    return html
season = 2016 

url = f"https://www.basketball-reference.com/leagues/NBA_{season}_games.html"
url
'https://www.basketball-reference.com/leagues/NBA_2016_games.html'

****html = await get_html(url, "#content .filter")**
Task exception was never retrieved
future: <Task finished name='Task-5' coro=<Connection.run() done, defined at C:\Users\Anthony Ford\AppData\Local\Programs\Python\Python311\Lib\site-packages\playwright\_impl\_connection.py:240> exception=NotImplementedError()>
Traceback (most recent call last):
  File "C:\Users\Anthony Ford\AppData\Local\Programs\Python\Python311\Lib\site-packages\playwright\_impl\_connection.py", line 247, in run
    await self._transport.connect()
  File "C:\Users\Anthony Ford\AppData\Local\Programs\Python\Python311\Lib\site-packages\playwright\_impl\_transport.py", line 132, in connect
    raise exc
  File "C:\Users\Anthony Ford\AppData\Local\Programs\Python\Python311\Lib\site-packages\playwright\_impl\_transport.py", line 120, in connect
    self._proc = await asyncio.create_subprocess_exec(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Anthony Ford\AppData\Local\Programs\Python\Python311\Lib\asyncio\subprocess.py", line 218, in create_subprocess_exec
    transport, protocol = await loop.subprocess_exec(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Anthony Ford\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py", line 1688, in subprocess_exec
    transport = await self._make_subprocess_transport(
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Anthony Ford\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py", line 502, in _make_subprocess_transport
    raise NotImplementedError**
**NotImplementedError
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
Cell In[12], line 1
----> 1 html = await get_html(url, "#content .filter")

Cell In[9], line 7, in get_html(url, selector, sleep, retries)
      4 time.sleep(sleep * i)
      6 try:
----> 7     async with async_playwright() as p:
      8         browser = await p.chromium.launch()
      9         page= await browser.new_page()

File ~\AppData\Local\Programs\Python\Python311\Lib\site-packages\playwright\async_api\_context_manager.py:46, in PlaywrightContextManager.__aenter__(self)
     44 if not playwright_future.done():
     45     playwright_future.cancel()
---> 46 playwright = AsyncPlaywright(next(iter(done)).result())
     47 playwright.stop = self.__aexit__  # type: ignore
     48 return playwright

File ~\AppData\Local\Programs\Python\Python311\Lib\site-packages\playwright\_impl\_transport.py:120, in PipeTransport.connect(self)
    117     if getattr(sys, "frozen", False):
    118         env.setdefault("PLAYWRIGHT_BROWSERS_PATH", "0")
--> 120     self._proc = await asyncio.create_subprocess_exec(
    121         str(self._driver_executable),
    122         "run-driver",
    123         stdin=asyncio.subprocess.PIPE,**
    124         stdout=asyncio.subprocess.PIPE,
    125         stderr=_get_stderr_fileno(),
    126         limit=32768,
    127         creationflags=creationflags,
    128         env=env,
    129     )
    130 except Exception as exc:
    131     self.on_error_future.set_exception(exc)

File ~\AppData\Local\Programs\Python\Python311\Lib\asyncio\subprocess.py:218, in create_subprocess_exec(program, stdin, stdout, stderr, limit, *args, **kwds)
    215 loop = events.get_running_loop()
    216 protocol_factory = lambda: SubprocessStreamProtocol(limit=limit,
    217                                                     loop=loop)
--> 218 transport, protocol = await loop.subprocess_exec(
    219     protocol_factory,
    220     program, *args,
    221     stdin=stdin, stdout=stdout,
    222     stderr=stderr, **kwds)
    223 return Process(transport, protocol, loop)

File ~\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py:1688, in BaseEventLoop.subprocess_exec(self, protocol_factory, program, stdin, stdout, stderr, universal_newlines, shell, bufsize, encoding, errors, text, *args, **kwargs)
   1686     debug_log = f'execute program {program!r}'
   1687     self._log_subprocess(debug_log, stdin, stdout, stderr)
-> 1688 transport = await self._make_subprocess_transport(
   1689     protocol, popen_args, False, stdin, stdout, stderr,
   1690     bufsize, **kwargs)
   1691 if self._debug and debug_log is not None:
   1692     logger.info('%s: %r', debug_log, transport)

File ~\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py:502, in BaseEventLoop._make_subprocess_transport(self, protocol, args, shell, stdin, stdout, stderr, bufsize, extra, **kwargs)
    498 async def _make_subprocess_transport(self, protocol, args, shell,
    499                                      stdin, stdout, stderr, bufsize,
    500                                      extra=None, **kwargs):
    501     """Create subprocess transport."""
--> 502     raise NotImplementedError

NotImplementedError: 

ht.sync_api import sync_playwright
with sync_playwright() as p:
    browser = p.chromium.launch()
    page = browser.new_page()
    # ...
    browser.close()
```

**Comment by mxschmitt** [comment]
See in #480, #462 and #178 for context and https://playwright.dev/python/docs/library#incompatible-with-selectoreventloop-of-asyncio-on-windows.

TL;DR: You need to use the async API.

**Comment by victorncg** [comment]
> Hello
> 
> Here is the code that i am working on and I keep getting this NotImplementedError: from the playwrite. Can someone help!!
> 
> ```
> import os
> from bs4 import BeautifulSoup
> from playwright.async_api import async_playwright, TimeoutError as PlaywrightTimeout
> import time
> %pip install pandas
> Requirement already satisfied: pandas in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (1.5.2)
> Requirement already satisfied: python-dateutil>=2.8.1 in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (from pandas) (2.8.2)
> Requirement already satisfied: pytz>=2020.1 in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (from pandas) (2022.7)
> Requirement already satisfied: numpy>=1.21.0 in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (from pandas) (1.24.1)
> Requirement already satisfied: six>=1.5 in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (from python-dateutil>=2.8.1->pandas) (1.16.0)
> Note: you may need to restart the kernel to use updated packages.
> %pip install beautifulsoup4
> Requirement already satisfied: beautifulsoup4 in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (4.11.1)
> Requirement already satisfied: soupsieve>1.2 in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (from beautifulsoup4) (2.3.2.post1)
> Note: you may need to restart the kernel to use updated packages.
> %pip install playwright
> Requirement already satisfied: playwright in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (1.29.1)
> Requirement already satisfied: greenlet==2.0.1 in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (from playwright) (2.0.1)
> Requirement already satisfied: pyee==9.0.4 in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (from playwright) (9.0.4)
> Requirement already satisfied: typing-extensions in c:\users\anthony ford\appdata\local\programs\python\python311\lib\site-packages (from pyee==9.0.4->playwright) (4.4.0)
> Note: you may need to restart the kernel to use updated packages.
> !playwright install
> SEASONS = list(range(2016, 2023))
> SEASONS
> [2016, 2017, 2018, 2019, 2020, 2021, 2022]
> DATA_DIR = "data"
> STANDINGS_DIR = os.path.join(DATA_DIR, "standings")
> SCORES_DIR = os.path.join(DATA_DIR, "scores")
> async def get_html(url, selector, sleep=5, retries=5):
>     html = None
>     for i in range(1, retries+1):
>         time.sleep(sleep * i)
>         
>         try:
>             async with async_playwright() as p:
>                 browser = await p.chromium.launch()
>                 page= await browser.new_page()
>                 await page.goto(url)
>                 print(await page.title())
>                 html = await page.inner_html(selector)
>         except PlaywrightTimeout:
>             print(f"Timeout error on {url}")
>             continue
>         else:
>             break
>     return html
> season = 2016 
> 
> url = f"https://www.basketball-reference.com/leagues/NBA_{season}_games.html"
> url
> 'https://www.basketball-reference.com/leagues/NBA_2016_games.html'
> 
> ****html = await get_html(url, "#content .filter")**
> Task exception was never retrieved
> future: <Task finished name='Task-5' coro=<Connection.run() done, defined at C:\Users\Anthony Ford\AppData\Local\Programs\Python\Python311\Lib\site-packages\playwright\_impl\_connection.py:240> exception=NotImplementedError()>
> Traceback (most recent call last):
>   File "C:\Users\Anthony Ford\AppData\Local\Programs\Python\Python311\Lib\site-packages\playwright\_impl\_connection.py", line 247, in run
>     await self._transport.connect()
>   File "C:\Users\Anthony Ford\AppData\Local\Programs\Python\Python311\Lib\site-packages\playwright\_impl\_transport.py", line 132, in connect
>     raise exc
>   File "C:\Users\Anthony Ford\AppData\Local\Programs\Python\Python311\Lib\site-packages\playwright\_impl\_transport.py", line 120, in connect
>     self._proc = await asyncio.create_subprocess_exec(
>                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\Anthony Ford\AppData\Local\Programs\Python\Python311\Lib\asyncio\subprocess.py", line 218, in create_subprocess_exec
>     transport, protocol = await loop.subprocess_exec(
>                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\Anthony Ford\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py", line 1688, in subprocess_exec
>     transport = await self._make_subprocess_transport(
>                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\Anthony Ford\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py", line 502, in _make_subprocess_transport
>     raise NotImplementedError**
> **NotImplementedError
> ---------------------------------------------------------------------------
> NotImplementedError                       Traceback (most recent call last)
> Cell In[12], line 1
> ----> 1 html = await get_html(url, "#content .filter")
> 
> Cell In[9], line 7, in get_html(url, selector, sleep, retries)
>       4 time.sleep(sleep * i)
>       6 try:
> ----> 7     async with async_playwright() as p:
>       8         browser = await p.chromium.launch()
>       9         page= await browser.new_page()
> 
> File ~\AppData\Local\Programs\Python\Python311\Lib\site-packages\playwright\async_api\_context_manager.py:46, in PlaywrightContextManager.__aenter__(self)
>      44 if not playwright_future.done():
>      45     playwright_future.cancel()
> ---> 46 playwright = AsyncPlaywright(next(iter(done)).result())
>      47 playwright.stop = self.__aexit__  # type: ignore
>      48 return playwright
> 
> File ~\AppData\Local\Programs\Python\Python311\Lib\site-packages\playwright\_impl\_transport.py:120, in PipeTransport.connect(self)
>     117     if getattr(sys, "frozen", False):
>     118         env.setdefault("PLAYWRIGHT_BROWSERS_PATH", "0")
> --> 120     self._proc = await asyncio.create_subprocess_exec(
>     121         str(self._driver_executable),
>     122         "run-driver",
>     123         stdin=asyncio.subprocess.PIPE,**
>     124         stdout=asyncio.subprocess.PIPE,
>     125         stderr=_get_stderr_fileno(),
>     126         limit=32768,
>     127         creationflags=creationflags,
>     128         env=env,
>     129     )
>     130 except Exception as exc:
>     131     self.on_error_future.set_exception(exc)
> 
> File ~\AppData\Local\Programs\Python\Python311\Lib\asyncio\subprocess.py:218, in create_subprocess_exec(program, stdin, stdout, stderr, limit, *args, **kwds)
>     215 loop = events.get_running_loop()
>     216 protocol_factory = lambda: SubprocessStreamProtocol(limit=limit,
>     217                                                     loop=loop)
> --> 218 transport, protocol = await loop.subprocess_exec(
>     219     protocol_factory,
>     220     program, *args,
>     221     stdin=stdin, stdout=stdout,
>     222     stderr=stderr, **kwds)
>     223 return Process(transport, protocol, loop)
> 
> File ~\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py:1688, in BaseEventLoop.subprocess_exec(self, protocol_factory, program, stdin, stdout, stderr, universal_newlines, shell, bufsize, encoding, errors, text, *args, **kwargs)
>    1686     debug_log = f'execute program {program!r}'
>    1687     self._log_subprocess(debug_log, stdin, stdout, stderr)
> -> 1688 transport = await self._make_subprocess_transport(
>    1689     protocol, popen_args, False, stdin, stdout, stderr,
>    1690     bufsize, **kwargs)
>    1691 if self._debug and debug_log is not None:
>    1692     logger.info('%s: %r', debug_log, transport)
> 
> File ~\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py:502, in BaseEventLoop._make_subprocess_transport(self, protocol, args, shell, stdin, stdout, stderr, bufsize, extra, **kwargs)
>     498 async def _make_subprocess_transport(self, protocol, args, shell,
>     499                                      stdin, stdout, stderr, bufsize,
>     500                                      extra=None, **kwargs):
>     501     """Create subprocess transport."""
> --> 502     raise NotImplementedError
> 
> NotImplementedError: 
> 
> ht.sync_api import sync_playwright
> with sync_playwright() as p:
>     browser = p.chromium.launch()
>     page = browser.new_page()
>     # ...
>     browser.close()
> ```

It doesn't work for me either. I keep getting the same result with my slightly altered code:

`from playwright.async_api import async_playwright

playwright = await async_playwright().start()
browser = await playwright.chromium.launch(headless = False)
page = await browser.new_page()

await page.goto("http://whatsmyuseragent.org/")`

I've read this thread and we might be having issues with the "event loop policy" of the jupyter kernel:https://stackoverflow.com/questions/71228742/how-to-use-the-playwright-library-in-a-jupyter-notebook-instead-of-using-a-regul

Haven't tried changing this yet, so let's see.
